### PR TITLE
Update the approach we use to layer classifications

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Idea though of by @sharwell .

This replaces the current approach we have with classification where we have 3 independent taggers (one for syntax, one for semantics, and one for embedded classification), and replaces it with 1 tagger that internally has 3 tagging components it defers to and aggregates.

THe problem with the former approach is that it led to undesirable scenarios where users might do things like comment out a line of code and then see semantic classifications still show up.  This was because each classifier was independent, and so the same section of code might classified as a comment by the syntactic classifier, but a symbol by the semantic one (which hadn't caught up yet).

Prior approaches to try to resolve this worked by attempting to place these classifications on different editor layers, which allowed things like comments to override non-commented code.  While this worked, Sam had a cleaner idea.  Instead of having different taggers that conflicted with each other, we would just have a single tagger that aggregated the results of all three sub-tagging-components.  This aggregating tagger could then add the smarts about having syntactic comments override semantic tags, as well as a nice way of having embedded tags (regex/json/c#-test) always override string tags.

This turned out to be nice and easy to do, and decouples us from teh editor layering system.  It also means we should no longer have to worry about any issues in the future stemming from our classifiers being independent, which may be nice in the future for future classification enhancements we may have.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs [hunk 1]
```diff
@@ -2714,7 +2714,7 @@ public async Task TestCreateWithBufferNotInWorkspace()
                 visibilityTracker: null,
                 listenerProvider);
 
-            using var tagger = (IDisposable?)provider.CreateTagger<IClassificationTag>(disposableView.TextView, extraBuffer);
+            using var tagger = provider.CreateTagger(disposableView.TextView, extraBuffer);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs:2717]: made things more strongly typed, which had the added benefit of removing hte need for speculative casting in tests.

File: src/EditorFeatures/CSharpTest/Classification/SyntacticTaggerTests.cs [hunk 2]
```diff
@@ -102,7 +102,7 @@ public async Task TestTagsChangedAfterDelete()
 
             var checkpoint = new Checkpoint();
 
-            var typeMap = workspace.ExportProvider.GetExportedValue<SyntacticClassificationTypeMap>();
+            var typeMap = workspace.ExportProvider.GetExportedValue<ClassificationTypeMap>();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Classification/SyntacticTaggerTests.cs:105]: got rid of the two type maps.  we only have one type map we use uniformly in roslyn now.

File: src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.TagComputer.cs [hunk 3]
```diff
@@ -358,17 +358,17 @@ async ValueTask<SnapshotSpan> ComputeChangedSpanAsync()
                     return _lastProcessedData;
             }
 
-            public IEnumerable<ITagSpan<IClassificationTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+            public SegmentedList<ITagSpan<IClassificationTag>> GetTags(NormalizedSnapshotSpanCollection spans)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.TagComputer.cs:361]: made things more strongly typed.  Allows for the consumption side (still within roslyn), to do things like non-allocate enumerate through things.

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:1]: this file is the meat of hte change.

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+{
+    private readonly SyntacticClassificationTaggerProvider _syntacticTaggerProvider = new(threadingContext, typeMap, globalOptions, listenerProvider);
+    private readonly SemanticClassificationViewTaggerProvider _semanticTaggerProvider = new(threadingContext, typeMap, globalOptions, visibilityTracker, listenerProvider);
+    private readonly EmbeddedLanguageClassificationViewTaggerProvider _embeddedTaggerProvider = new(threadingContext, typeMap, globalOptions, visibilityTracker, listenerProvider);
+
+    public ITagger<T>? CreateTagger<T>(ITextView textView, ITextBuffer buffer) where T : ITag
+    {
+        var syntacticTagger = _syntacticTaggerProvider.CreateTagger(buffer);
+        var semanticTagger = _semanticTaggerProvider.CreateTagger(textView, buffer);
+        var embeddedTagger = _embeddedTaggerProvider.CreateTagger(textView, buffer);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:46]: making things strongly typed meant a lot less checking for idisposable, and needing to do things like `ITagger<T>` tests on these as well.

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+            var totalTags = new SegmentedList<ITagSpan<IClassificationTag>>();
+
+            using var _ = Classifier.GetPooledList<ITagSpan<IClassificationTag>>(out var stringLiterals);
+
+            var syntacticSpans = syntacticTagger.GetTags(spans).GetEnumerator();
+            var semanticSpans = semanticTagger.GetTags(spans).GetEnumerator();
+
+            var currentSyntactic = NextOrNull(ref syntacticSpans);
+            var currentSemantic = NextOrNull(ref semanticSpans);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:89]: this follows a similar pattern we use when we produce tags and diff them against the last produced tags.

File: src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs [hunk 7]
```diff
@@ -2691,40 +2693,6 @@ void goo()
                 Method("nameof"));
         }
 
-        [WpfFact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/744813")]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs:2694]: just moved to bottom of file.

File: src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs [hunk 8]
```diff
@@ -3935,5 +3903,117 @@ void M(Action action)
...
+            }
+
+            var waiter = listenerProvider.GetWaiter(FeatureAttribute.Classification);
+            await waiter.ExpeditedWaitAsync();
+        }
+
+        [WpfFact]
+        public async Task TestTotalClassifier()
+        {
+            // don't crash
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs:3944]:
```suggestion
```

File: src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs [hunk 9]
```diff
@@ -3935,5 +3903,116 @@ void M(Action action)
                 Method("staticLocalFunction"),
                 Static("staticLocalFunction"));
         }
+
+        [WpfFact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/744813")]
+        public async Task TestCreateWithBufferNotInWorkspace()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs:3908]: validates that we get syntactic/semantic/embedded from the one tagger.

File: src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs [hunk 10]
```diff
@@ -3935,5 +3903,116 @@ void M(Action action)
...
+                edit.Insert(0, "class A { }");
+                edit.Apply();
+            }
+
+            var waiter = listenerProvider.GetWaiter(FeatureAttribute.Classification);
+            await waiter.ExpeditedWaitAsync();
+        }
+
+        [WpfFact]
+        public async Task TestTotalClassifier()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Classification/SemanticClassifierTests.cs:3942]: validates that we get syntactic/semantic/embedded from the one tagger.

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+            syntacticSpans.Sort(s_spanComparison);
+            semanticSpans.Sort(s_spanComparison);
+
+            var syntacticEnumerator = syntacticSpans.GetEnumerator();
+            var semanticEnumerator = semanticSpans.GetEnumerator();
+
+            var currentSyntactic = GetNextSyntacticSpan();
+            var currentSemantic = GetNextSemanticSpan();
+
+            while (currentSyntactic != null && currentSemantic != null)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:102]: this follows a similar strategy/code we have when producing tags where we diff the old and new tags against each other.  because they're sorted, we can just walk them in order, taking from the earlier one to produce the final result.  one when we hit spns of interest do we have to look at the other and decide what to do.

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+                    return false;
+
+                // Keep skipping semantic tags that overlaps with this syntactic tag.
+                while (currentSemantic != null && currentSemantic.Span.OverlapsWith(currentSyntactic.Span.Span))
+                    currentSemantic = GetNextSemanticSpan();
+
+                // now add that syntactic span.
+                totalTags.Add(currentSyntactic);
+                currentSyntactic = GetNextSyntacticSpan();
+
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:180]:
```suggestion
```

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+            var syntacticEnumerator = syntacticSpans.GetEnumerator();
+            var semanticEnumerator = semanticSpans.GetEnumerator();
+
+            var currentSyntactic = GetNextSyntacticSpan();
+            var currentSemantic = GetNextSemanticSpan();
+
+            while (currentSyntactic != null && currentSemantic != null)
+            {
+                // as long as we see semantic spans before the next syntactic one, keep adding them.
+                if (currentSemantic.Span.Start <= currentSyntactic.Span.Start)
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:105]: ><= [](http://example.com/codeflow?start=47&length=2)

Just making sure the == case isn't worth worrying about

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:105]: yeah, this is intentional.  if semantic starts on or before syntactic, we just take as is. 

we care about when syntactic is before, as syntactic can bleed into semantic based on where the start of the syntactic token in (for comments/strings/excluded code).

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Classification;
+
+[Export(typeof(IViewTaggerProvider))]
+[TagType(typeof(IClassificationTag))]
+[Microsoft.VisualStudio.Utilities.ContentType(ContentTypeNames.RoslynContentType)]
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:28]: >Microsoft.VisualStudio.Utilities. [](http://example.com/codeflow?start=1&length=33)

Are there types in this ns that cause conflicts?

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+[method: ImportingConstructor]
+[method: SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+internal sealed class TotalClassificationTaggerProvider(
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:31]: >Total [](http://example.com/codeflow?start=22&length=5)

nit: the "Total" part of the name seems a bit odd to me. Maybe "Complete" or "Aggregated"?

File: src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs [hunk 16]
```diff
@@ -2,43 +2,47 @@
...
         private readonly IAsynchronousOperationListener _listener = listenerProvider.GetListener(FeatureAttribute.Classification);
         private readonly IThreadingContext _threadingContext = threadingContext;
-        private readonly SyntacticClassificationTypeMap _typeMap = typeMap;
+        private readonly ClassificationTypeMap _typeMap = typeMap;
         private readonly IGlobalOptionService _globalOptions = globalOptions;
 
         private readonly ConditionalWeakTable<ITextBuffer, TagComputer> _tagComputers = new();
 
-        public ITagger<T>? CreateTagger<T>(ITextBuffer buffer) where T : ITag
+        ITagger<T>? ITaggerProvider.CreateTagger<T>(ITextBuffer buffer)
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs:33]: >ITagger<T>? ITaggerProvider.CreateTagger<T>(ITextBuffer buffer) [](http://example.com/codeflow?start=8&length=63)

Not part of this PR, but I'm curious why this uses a CWT instead of the buffer properties to store the TagComputer

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs (hunk 14 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:28]: yes :'(

File: src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs (hunk 16 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs:33]: that is a wonderful question...

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs:33]: @jasonmalinowski  do you recall any reason for this?  I'm happy to change this in a followup pr.

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+                // Nothing complex to do if we got no embedded classifications back.  Just add in all the string
+                // classifications, untouched.
+                if (embeddedClassifications.Count == 0)
+                {
+                    totalTags.AddRange(stringLiterals);
+                    return;
+                }
+
+                // ClassifierHelper.MergeParts requires these to be sorted.
+                stringLiterals.Sort(s_spanComparison);
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:202]: >stringLiterals.Sort(s_spanComparison); [](http://example.com/codeflow?start=16&length=38)

I guess being explicit is good, but isn't NormalizedSnapshotSpanCollection guaranteed ordered?

Comment (ToddGrun) [reply src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:202]: whoops, this is over the literals not the spans, never mind

File: src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+
+                // Call into the helper to merge the string literals and embedded classifications into the final result.
+                // The helper will add all the embedded classifications first, then add string literal classifications
+                // in the the space between the embedded classifications that were originally classified as a string
+                // literal.
+                ClassifierHelper.MergeParts<ITagSpan<IClassificationTag>, ClassificationTagSpanIntervalIntrospector>(
+                    stringLiterals,
+                    embeddedClassifications,
+                    totalTags,
+                    static tag => tag.Span.Span.ToTextSpan(),
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Classification/TotalClassificationTaggerProvider.cs:213]: >Span.Span.ToTextSpan [](http://example.com/codeflow?start=38&length=20)

I feel like you're trying to do something with a Span here :)

File: src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs (hunk 16 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs:33]: handlign in a followup PR.

File: src/EditorFeatures/Core/Tagging/AggregateTagger.cs [hunk 19]
```diff
@@ -5,49 +5,59 @@
...
+    : AbstractAggregateTagger<TTag>(taggers)
+    where TTag : ITag
+{
+    public override SegmentedList<ITagSpan<TTag>> GetTags(NormalizedSnapshotSpanCollection spans)
     {
-        using var _ = ArrayBuilder<ITagSpan<TTag>>.GetInstance(out var result);
+        var result = new SegmentedList<ITagSpan<TTag>>();
 
-        foreach (var tagger in _taggers)
+        foreach (var tagger in this.Taggers)
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/Tagging/AggregateTagger.cs:58]: >foreach (var tagger in this.Taggers) [](http://example.com/codeflow?start=8&length=36)

Probably not worth it, but if we're expecting this to get large (which is kind of implied by SegmentedList usage), would it be better to have the GetTags take in a builder that it can just directly inject into?

File: src/EditorFeatures/CSharpTest/Classification/TotalClassifierTests.cs [hunk 20]
```diff
@@ -3022,5 +3028,118 @@ public async Task TestFunctionPointer2(TestHost testHost)
...
+            var buffer = document.GetTextBuffer();
+            using var tagger = provider.CreateTagger(document.GetTextView(), buffer);
+
+            var waiter = listenerProvider.GetWaiter(FeatureAttribute.Classification);
+            await waiter.ExpeditedWaitAsync();
+
+            var allCode = buffer.CurrentSnapshot.GetText();
+            var tags = tagger!.GetTags(new NormalizedSnapshotSpanCollection(buffer.CurrentSnapshot.GetFullSpan()));
+
+            var actualOrdered = tags.OrderBy((t1, t2) => t1.Span.Span.Start - t2.Span.Span.Start);
```
Comment (ToddGrun) [inline src/EditorFeatures/CSharpTest/Classification/TotalClassifierTests.cs:3074]: >tags.OrderBy((t1, t2) => t1.Span.Span.Start - t2.Span.Span.Start); [](http://example.com/codeflow?start=32&length=66)

any concern about items with the same start being ordered inconsistently? (ie, should it also compare end, classificationtype to ensure consistent ordering?)

File: src/Workspaces/Core/Portable/Classification/Classifier.cs [hunk 21]
```diff
@@ -21,9 +21,12 @@ namespace Microsoft.CodeAnalysis.Classification
     public static class Classifier
     {
         internal static PooledObject<SegmentedList<ClassifiedSpan>> GetPooledList(out SegmentedList<ClassifiedSpan> classifiedSpans)
+            => GetPooledList<ClassifiedSpan>(out classifiedSpans);
+
+        internal static PooledObject<SegmentedList<T>> GetPooledList<T>(out SegmentedList<T> classifiedSpans)
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Classification/Classifier.cs:26]: >internal static PooledObject<SegmentedList<T>> GetPooledList<T>(out SegmentedList<T> classifiedSpans) [](http://example.com/codeflow?start=8&length=101)

This doesn't feel specific to ClassifiedSpans anymore. Would it make sense for this to be in a more general location?

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Core/Portable/Classification/Classifier.cs (hunk 21 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Classification/Classifier.cs:26]: yes.  will address in followup.

File: src/EditorFeatures/Core/Tagging/AggregateTagger.cs (hunk 19 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Tagging/AggregateTagger.cs:58]: yes.  will address in followup.

File: src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs (hunk 16 above)
Comment (jasonmalinowski) [reply src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.cs:33]: I believe the only reason for CWTs was because some folks were scared to use buffer properties because it means somebody could _theoretically_ grab the objects out of the property bags.
